Fix cluster entity behavior

tgraf opened the pull request:
Requires: #5771 (to solve upgrade impact)

This completely removes the cluster identity and removes almost all of the logic depending on being aware of the cluster CIDR range.

What currently remains is:
* The SNAT rule for traffic into `cilium_host` excludes the cluster range if direct routing is configured. This rule is harder to remove. A possibility is to maintain an ipset with all node prefixes and add/remove it based on node discovery.

It also fixes the behavior of the cluster entity to the following new definition:
 - All known endpoints in the local cluster, i.e. all endpoints which have a
   cluster label that equals to the cluster name associated with the
   local cilium-agent.
 - The host identity to include communication to cluster nodes and host
   networking endpoints
 - The init identity to allow communication to endpoints starting up in
   the local cluster.

Fixes: #5739

## TODO
- [x] Evaluate whether the addition of the cluster label must be hidden behind an opt-in flag
- [x] Upgrade test
- [x] Evaluate whether cluster entity should include host or whether two entities should be used to define endpoints only and endpoints+host

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5581)
<!-- Reviewable:end -->

tgraf committed:
bpf: Remove cluster identity

This commit removes all logic that involves associating the cluster range with
the cluster identity. The cluster range cannot be detected automatically in a
reliable manner which led to undefined behavior with several side effects:

 * If the cluster range was assumed to be too broad, IPs outside of the cluster
   were mistakenly associated with the identity reserved:cluster.
 * The cluster identity was not a superset of all pods and worker node IPs which
   lead to the inability to accurately define the behavior allowing communication
   to or from the cluster identity.
 * CIDR rules apply to IP addresses that were not associated with the cluster
   identity. If the detection of the cluster range was not accurate, CIDR
   rules would in turn be incorrect as well.

In order to resolve the above side effects, the following changes are
performed:

1. Egress traffic that cannot be associated with a particular identity is no
   longer associated with the identity cluster even if the IP is within the
   assumed cluster range. Any IP address that is not associated with an
   endpoint or a host via the ipcache is now classified as world.

2. The detection of mapped IPv6 addresses no longer depends on the cluster
   identity. This was merely an optimization and is unnecessary.

3. For packets received via cilium_host, the node prefix based lookup to find
   the encapsulation target is now performed for all destinations. Regardless
   of whether the destination IP is within the cluster prefix or not. This was
   already the slow path as the majority of traffic via the cilium_host
   interface is targeted to local endpoints. Even for non-local traffic, it
   is an exception if the destination IP cannot be found in the ipcache to
   derive the encapsulation target. The small number of additional lookups is
   an extreme corner case.

4. All pod IP to hostIP mappings as learned via the Kubernetes pod
   notifications before the kvstore is available will now have the identity
   init instead of cluster. The documentation states that prior to kvstore
   connectivity, only fixed and well-known identities are reliable for policy
   enforcement purposes so this change is tolerable.

5. The cluster identity is removed from all CIDR based identities.

Upgrade impact:

1. Any eventual existing policy to or from reserved:cluster will become
   ineffective. The cluster identity was never documented directly in the
   context of policy enforcement so usage should be minimal to non-existent.
   The behavior was undefined to the most extend as cluster range detection is
   inaccurate. Continued use of reserved:cluster for policy is discouraged.

   However: the cluster entity will be redefined in a follow-up commit to
   provide a reliable mechanism to allow all traffic within the cluster.

2. Endpoint being restored with cluster identities whitelisted will be updated
   as the policy for those endpoints is recalculated for the first time. The
   regeneration on restore will automatically remove the logic to derive any
   traffic to the cluster identity.

3. Existing CIDR identities in the ipcache/allocator can continue to include
   the cluster range prefix. As agents are being restarted in a cluster, new
   CIDR identities will be allocated without the cluster range prefix and the
   old identities are phased out. No connectivity outage is expected from this.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Don't rely on cluster-range for ACCEPT rule for FORWARD chain

k8s < 1.8 requires a workaround as kube-proxy does not change the default
policy of the FORWARD chain. The existing ACCEPT rules depend on the
cluster range to avoid an allow-all policy on the FORWARD chain. In order
to no longer depend on the cluster range, the rules can be changed
as follows:

* The egress rule can be made more restrictive to only allow packets which
  originate from the local allocation range. This avoids depending on
  the cluster range and improves the security.

* The ingress rule can be made more liberal as it is already tied to the
  route which points to the cilium_host interface. This means that traffic is
  already limited to the set of all node allocation ranges.

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer commented on pkg/labels/cidr/cidr_test.go:
```diff
@@ -121,7 +109,7 @@ func (s *CIDRLabelsSuite) TestGetCIDRLabelsInCluster(c *C) {
 		"cidr:0.0.0.0/0",
 		"cidr:10.0.0.0/16",
-		"reserved:cluster",
+		"reserved:world",
```
We can probably drop this entire test, there's no difference between in-cluster / out-of-cluster CIDR identities any more.

joestringer reviewed (dismissed):
This looks worthy of a release note.

Also wondering about the upgrade considerations:

* These labels may exist in kvstore (eg in existing labels->identity mappings for CIDRs in the cluster prefix). I suppose that newer Cilium instances will generate different identities for the same CIDR and not use the ones that would previously be used. ~~During upgrade this could cause some disruption as newer Cilium instances roll out using the new identity and older instances continue to operate on the old identity for the same CIDR.~~ Given that these would only be used on egress from a given node, the inconsistency between nodes shouldn't cause problems with connectivity between nodes.

* We still have the entity cluster, but I can't quite figure out what it will do. We presumably would need this to avoid parse errors when importing existing CNPs during upgrade (otherwise we would reject the policy). Does it just map to an identity that never matches anything?

How can we make it obvious to users that they are using an entity in their policies that no longer works?

tgraf committed:
policy: Allow an entity to map to EndpointSelectorSlice

This commit does not change the code functionality. It converts the existing definition
from Entity -> EndpointSelector to Entity -> EndpointSelectorSlice.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
policy: Define well-known endpoint selectors

This commit introduces no functional changes. It prepares endpoint selectors to be
used in multiple entity definitions.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
test: Runtime policy test for entity "all" to allow to any external IP

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
policy: Fix behavior of cluster entity

The cluster entity has been somewhat undefined in the past. It consisted of IPs
in the cluster IP range that could not be associated with a more specific
identity. This rendered the use of the entity entirely useless.

Re-define the behavior of the cluster entity to mean the following:
 - All known endpoints in the local cluster, i.e. all endpoints which have a
   cluster label that equals to the cluster name associated with the
   local cilium-agent.
 - The host identity to include communication to cluster nodes and host
   networking endpoints
 - The init identity to allow communication to endpoints starting up in
   the local cluster.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
endpoint: Always include cluster label in endpoint

The cluster label has been included with the endpoint only when the cluster
name was set to a non-default value. Always include the cluster name to allow
using it to implement the cluster policy entity.

Upgrade impact:
This adds an additional label to all endpoints as the endpoint is being
restored. This has the potential to introduce connectivity outages while the
cluster adjusts to the new security identity allocated to the endpoint. The
upgrade impact is being addressed in PR #5771

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer commented on pkg/policy/api/entity.go:
```diff
@@ -44,37 +44,49 @@ const (
-	if selector, ok := EntitySelectorMapping[e]; ok {
-		return selector.Matches(ctx)
+	if selectors, ok := EntitySelectorMapping[e]; ok {
+		for _, selector := range selectors {
```
Open-coding `EndpointSelectorSlice.Matches()`?

joestringer reviewed (commented):
Do you think there might be a case where someone wants to allow ingress from all endpoints in the cluster but not from nodes in the underlay? Wouldn't it be more explicit to require users to also specify host in their policies?

tgraf committed:
doc: Fix and improve entity documentation

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Document all reserved identities

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
test: Add k8s runtime tests for toEntities { all | world | cluster | host }

The host test is only a negative test for now.

Signed-off-by: Thomas Graf <thomas@cilium.io>

jrajahalme reviewed (approved):
Approving changes to pkg/proxy/logger/logger.go on behalf of the "proxy" area.

raybejjani commented on bpf/lib/policy.h:
```diff
@@ -30,7 +30,6 @@
  * - IdentityUnknown		(0)
  * - ReservedIdentityHost	(1)
  * - ReservedIdentityWorld	(2)
- * - ReservedIdentityCluster	(3)
```
It might be better to leave this marked in some way? I would hate to re-use this identity and forget that it meant something to older cilium versions (same for the actual define in node_config.h)

raybejjani commented on pkg/defaults/node.go:
```diff
@@ -25,9 +25,6 @@ const (
 	IPv6NodePrefixLen = 96
 
-	// DefaultIPv6ClusterPrefixLen is the IPv6 prefix length of the entire cluster.
-	DefaultIPv6ClusterPrefixLen = 64
```
I haven't checked, but should `DefaultIPv4ClusterPrefixLen` also be removed?

tgraf replied:
There is one use remaining right now:
```
                ingressSnatSrcAddrExclusion := node.GetHostMasqueradeIPv4().String()
                if option.Config.Tunnel == option.TunnelDisabled {
                        ingressSnatSrcAddrExclusion = node.GetIPv4ClusterRange().String()
                }
```

raybejjani commented on daemon/daemon.go:
```diff
@@ -550,7 +549,7 @@ func (d *Daemon) installIptablesRules() error {
 	if err := runProg("iptables", []string{
 		"-A", ciliumForwardChain,
-		"-s", node.GetIPv4ClusterRange().String(),
+		"-s", node.GetIPv4AllocRange().String(),
```
The comment above may need to be updated? `GetIPv4ClusterRange` is the range for this node, and that might change this behaviour? I'm not sure if it was intended to be specific to the node range, or a catch-all for any packet sent to the overlay.

raybejjani commented on daemon/daemon.go:
```diff
@@ -539,7 +539,6 @@ func (d *Daemon) installIptablesRules() error {
 	if err := runProg("iptables", []string{
 		"-A", ciliumForwardChain,
-		"-d", node.GetIPv4ClusterRange().String(),
 		"-o", "cilium_host",
```
Wouldn't `GetIPv4AllocRange` go here, like the rule below?

tgraf replied:
Unfortunately, that will be too restrictive as a local process talking to a remote pod IP will go this path and will need to be masqueraded. As written in the commit message, it is safe to ignore the destination IP as the path into cilium_host can only be taken if a route is pointing to it. I've clarified this with a better comment.

raybejjani reviewed (dismissed):
We seem to remove some of the IPv4 stuff in some places, and IPv6 in others. I can probably dig in more but it what's the distinction? I was expecting the same changes to IPv4 and IPv6.

ianvernon commented on bpf/bpf_lxc.c:
```diff
@@ -117,7 +117,6 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 				   __u32 *dstID)
```
Typos in commit: 

"is not n endpoint identity or a host" --> "is not an endpoint identity or a host"
"The behavior was undefined to the most extend as cluster range detection is
   inaccurate." - not sure what this means.
"When the ipcache is first populated based on pod notifications, before the
   kvstore is available." - not a complete sentence.

ianvernon commented on Documentation/policy/language.rst:
```diff
@@ -292,12 +292,25 @@ by the selected endpoints.
+    containers running in host networking mode.
+cluster
+    Cluster is the logical group of all network endpoints inside of the local
+    cluster. This includes all Cilium managed endpoints of the local cluster.
```
Cilium-managed

ianvernon commented on Documentation/concepts.rst:
```diff
@@ -380,11 +380,24 @@ identities are prefixed with the string ``reserved:``.
+| reserved:health     | This is health checking traffic generated by      |
+|                     | Cilium agents.                                    |
++---------------------+---------------------------------------------------+
+| reserved:init       | An endpoint for which the identity has not be     |
```
... for which the identity has not yet been resolved is assigned the init identity.

ianvernon commented on Documentation/concepts.rst:
```diff
@@ -380,11 +380,24 @@ identities are prefixed with the string ``reserved:``.
+|                     | resolved for yet is assigned the init identity.   |
+|                     | This represents the phase of an endpoint in which |
+|                     | some of the metadata required to derive the       |
+|                     | security identity is still missing.               |
```
Maybe note that this is most common in the case of endpoint bootstrapping?

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -131,6 +135,72 @@ var _ = Describe("K8sPolicyTest", func() {
+					helpers.CurlFail("http://www.google.com/"))
+
+				if expectWorldSuccess {
+					res.ExpectSuccess("Egress connectivity to google.com should be allowed for pod %q", pod)
```
ExpectWithOffset? Similar for other `Expect*` in this function.

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -131,6 +135,72 @@ var _ = Describe("K8sPolicyTest", func() {
+				res := kubectl.ExecPodCmd(
+					helpers.DefaultNamespace, pod,
+					helpers.CurlFail("http://www.google.com/"))
+
```
Would be helpful to add calls to `By` here that contain the values of `expectWorldSuccess` and `expectClusterSuccess`. It will make the test more readable / easier to follow.
